File: crates/types/src/provider/request_record.rs
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

const DEFAULT_REQUEST_RECORD_LIMIT: u64 = 20;

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
pub struct RequestRecordListRequest {
    #[serde(default)]
    pub skip: u64,
    #[serde(default = "default_request_record_limit")]
    pub limit: u64,
    #[serde(default)]
    pub search: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub model_id: Option<String>,
    #[serde(default)]
    pub provider_id: Option<String>,
    #[serde(default)]
    pub api_format: Option<String>,
    #[serde(default, rename = "type")]
    pub type_filter: Option<String>,
}

impl Default for RequestRecordListRequest {
    fn default() -> Self {
        Self {
            skip: 0,
            limit: default_request_record_limit(),
            search: None,
            status: None,
            model_id: None,
            provider_id: None,
            api_format: None,
            type_filter: None,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RequestRecordListResponse {
    pub records: Vec<RequestRecord>,
    pub total: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
pub struct ActiveRequestRecordRequest {
    #[serde(default)]
    pub ids: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct ActiveRequestRecordResponse {
    pub records: Vec<RequestRecord>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RequestRecord {
    pub request_id: String,
    pub created_at: String,
    pub user_id: Option<String>,
    pub username: Option<String>,
    pub token_id: Option<String>,
    pub token_name: Option<String>,
    pub token_prefix: Option<String>,
    pub group_code: Option<String>,
    pub global_model_id: Option<String>,
    pub model_name: Option<String>,
    pub provider_id: Option<String>,
    pub provider_name: Option<String>,
    pub provider_key_name: Option<String>,
    pub provider_key_preview: Option<String>,
    pub client_api_format: String,
    pub provider_api_format: Option<String>,
    pub request_type: String,
    pub is_stream: bool,
    pub has_failover: bool,
    pub has_retry: bool,
    pub status: String,
    pub billing_status: String,
    pub client_status_code: Option<i32>,
    pub client_error_type: Option<String>,
    pub client_error_message: Option<String>,
    pub termination_origin: Option<String>,
    pub termination_reason: Option<String>,
    pub stream_end_reason: Option<String>,
    pub prompt_tokens: Option<i64>,
    pub completion_tokens: Option<i64>,
    pub total_tokens: Option<i64>,
    pub cache_creation_input_tokens: Option<i64>,
    pub cache_read_input_tokens: Option<i64>,
    #[serde(with = "rust_decimal::serde::float")]
    pub total_cost: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub token_cost: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub base_cost: Decimal,
    #[serde(with = "rust_decimal::serde::float")]
    pub billing_multiplier: Decimal,
    pub cost_currency: String,
    pub first_byte_time_ms: Option<i64>,
    pub total_latency_ms: Option<i64>,
    pub candidate_count: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RequestRecordDetail {
    pub record: RequestRecord,
    pub candidates: Vec<RequestCandidateDetail>,
    pub request_headers: Option<serde_json::Value>,
    pub request_body: Option<serde_json::Value>,
    pub client_response_headers: Option<serde_json::Value>,
    pub client_response_body: Option<serde_json::Value>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct RequestCandidateDetail {
    pub id: String,
    pub request_id: String,
    pub provider_id: Option<String>,
    pub provider_name: Option<String>,
    pub endpoint_id: Option<String>,
    pub endpoint_name: Option<String>,
    pub key_id: Option<String>,
    pub key_name: Option<String>,
    pub key_preview: Option<String>,
    pub client_api_format: String,
    pub provider_api_format: Option<String>,
    pub needs_conversion: bool,
    pub is_stream: bool,
    pub candidate_index: i32,
    pub retry_index: i32,
    pub status: String,
    pub skip_reason: Option<String>,
    pub status_code: Option<i32>,
    pub prompt_tokens: Option<i64>,
    pub completion_tokens: Option<i64>,
    pub total_tokens: Option<i64>,
    pub cache_creation_input_tokens: Option<i64>,
    pub cache_read_input_tokens: Option<i64>,
    #[serde(with = "rust_decimal::serde::float_option")]
    pub token_cost: Option<Decimal>,
    #[serde(with = "rust_decimal::serde::float_option")]
    pub base_cost: Option<Decimal>,
    #[serde(with = "rust_decimal::serde::float_option")]
    pub total_cost: Option<Decimal>,
    #[serde(with = "rust_decimal::serde::float_option")]
    pub billing_multiplier: Option<Decimal>,
    pub cost_currency: Option<String>,
    pub latency_ms: Option<i64>,
    pub first_byte_time_ms: Option<i64>,
    pub error_type: Option<String>,
    pub error_message: Option<String>,
    pub error_code: Option<String>,
    pub error_param: Option<String>,
    pub provider_request_headers: Option<serde_json::Value>,
    pub provider_request_body: Option<serde_json::Value>,
    pub provider_response_headers: Option<serde_json::Value>,
    pub provider_response_body: Option<serde_json::Value>,
    pub created_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
}

fn default_request_record_limit() -> u64 {
    DEFAULT_REQUEST_RECORD_LIMIT
}

File: apps/hook_frontend/src/types/request-record.ts
export type RequestRecordStatus = 'pending' | 'streaming' | 'success' | 'failed' | 'cancelled';

type RequestBillingFields = {
  service_tier?: string | null;
  input_cost?: number | null;
  output_cost?: number | null;
  cache_creation_cost?: number | null;
  cache_read_cost?: number | null;
  request_cost?: number | null;
  input_price_per_million?: number | null;
  output_price_per_million?: number | null;
  cache_creation_price_per_million?: number | null;
  cache_read_price_per_million?: number | null;
};

export type RequestRecord = RequestBillingFields & {
  request_id: string;
  created_at: string;
  user_id?: string | null;
  username?: string | null;
  token_id?: string | null;
  token_name?: string | null;
  token_prefix?: string | null;
  group_code?: string | null;
  global_model_id?: string | null;
  model_name?: string | null;
  provider_id?: string | null;
  provider_name?: string | null;
  provider_key_name?: string | null;
  provider_key_preview?: string | null;
  client_api_format: string;
  provider_api_format?: string | null;
  request_type: string;
  is_stream: boolean;
  has_failover: boolean;
  has_retry: boolean;
  status: RequestRecordStatus;
  billing_status: string;
  client_status_code?: number | null;
  client_error_type?: string | null;
  client_error_message?: string | null;
  termination_origin?: string | null;
  termination_reason?: string | null;
  stream_end_reason?: string | null;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens?: number | null;
  cache_creation_input_tokens?: number | null;
  cache_read_input_tokens?: number | null;
  input_text_tokens?: number | null;
  input_audio_tokens?: number | null;
  input_image_tokens?: number | null;
  output_text_tokens?: number | null;
  output_audio_tokens?: number | null;
  output_image_tokens?: number | null;
  reasoning_tokens?: number | null;
  cache_creation_5m_input_tokens?: number | null;
  cache_creation_1h_input_tokens?: number | null;
  usage_source?: string | null;
  usage_semantic?: string | null;
  total_cost: number;
  token_cost: number;
  base_cost: number;
  billing_multiplier: number;
  cost_currency: string;
  first_byte_time_ms?: number | null;
  total_latency_ms?: number | null;
  candidate_count: number;
};

export type RequestRecordListResponse = {
  records: RequestRecord[];
  total: number;
};

export type UsageRecord = {
  created_at: string;
  token_name?: string | null;
  token_prefix?: string | null;
  model_name?: string | null;
  client_api_format: string;
  request_type: string;
  is_stream: boolean;
  status: RequestRecordStatus;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens?: number | null;
  cache_creation_input_tokens?: number | null;
  cache_read_input_tokens?: number | null;
  total_cost: number;
  cost_currency: string;
  first_byte_time_ms?: number | null;
  total_latency_ms?: number | null;
};

export type UsageRecordListResponse = {
  records: UsageRecord[];
  total: number;
};

export type ActiveRequestRecordResponse = {
  records: RequestRecord[];
};

export type RequestCandidateDetail = RequestBillingFields & {
  id: string;
  request_id: string;
  provider_id?: string | null;
  provider_name?: string | null;
  endpoint_id?: string | null;
  endpoint_name?: string | null;
  key_id?: string | null;
  key_name?: string | null;
  key_preview?: string | null;
  client_api_format: string;
  provider_api_format?: string | null;
  needs_conversion: boolean;
  is_stream: boolean;
  candidate_index: number;
  retry_index: number;
  status: string;
  skip_reason?: string | null;
  status_code?: number | null;
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens?: number | null;
  cache_creation_input_tokens?: number | null;
  cache_read_input_tokens?: number | null;
  input_text_tokens?: number | null;
  input_audio_tokens?: number | null;
  input_image_tokens?: number | null;
  output_text_tokens?: number | null;
  output_audio_tokens?: number | null;
  output_image_tokens?: number | null;
  reasoning_tokens?: number | null;
  cache_creation_5m_input_tokens?: number | null;
  cache_creation_1h_input_tokens?: number | null;
  usage_source?: string | null;
  usage_semantic?: string | null;
  token_cost?: number | null;
  base_cost?: number | null;
  total_cost?: number | null;
  billing_multiplier?: number | null;
  cost_currency?: string | null;
  latency_ms?: number | null;
  first_byte_time_ms?: number | null;
  error_type?: string | null;
  error_message?: string | null;
  error_code?: string | null;
  error_param?: string | null;
  provider_request_headers?: unknown | null;
  provider_request_body?: unknown | null;
  provider_response_headers?: unknown | null;
  provider_response_body?: unknown | null;
  created_at: string;
  started_at?: string | null;
  finished_at?: string | null;
};

export type RequestRecordDetail = {
  record: RequestRecord;
  candidates: RequestCandidateDetail[];
  request_headers?: unknown | null;
  request_body?: unknown | null;
  client_response_headers?: unknown | null;
  client_response_body?: unknown | null;
};
